### PR TITLE
dropbear: use dropbearkey as ssh-keygen

### PR DESCRIPTION
**Pull request by stokito**
By default the OpenWrt is compiled with DropBear's clone of the OpenSSH `ssh-keygen` called [dropbearkey](https://manpages.ubuntu.com/manpages/noble/en/man1/dropbearkey.1.html). But it can be compiled with the OpenSSH client instead of the dropbear. 

Their commands differs that's why on the wiki page [Generating public and private keys](https://openwrt.org/docs/guide-user/security/dropbear.public-key.auth) we have two versions.

When developing luci-app-sshtunnel I added a button to generate a new key pair. And I had to add a [check](https://github.com/openwrt/luci/blob/master/applications/luci-app-sshtunnel/htdocs/luci-static/resources/view/sshtunnel/ssh_keys.js#L115C12-L115C12
): use the ssh-keygen by default but if it absent then use the dropbearkey.

Their commands also differs because flags `-q -N ''` wasn't supported by dropbearkey but this is fixed now. So we can now safely make an alias `ssh-keygen` to the `dropbearkey`.
Then users can use same commands in their scripts, have a better interoperability, reuse an old knowledge and not waste time to find a proper command to generate a cert with the dropbearkey.

The PR will create a symlink from `ssh-keygen` to `dropbear`. The DropBear should recognize the `ssh-keygen` programs and it's author is fine to add this support https://github.com/mkj/dropbear/issues/263

Then we'll need to add an alternative for the `openssh-keygen` so on installation it will substitute the DropBear's `ssh-keygen` with own. https://github.com/openwrt/packages/commit/9a3d317e151b77c63746a922806f2c7011d22668

Please confirm that you are fine with the change and will accept the PR. If yes, then I'll make a PR to the dropbear and openssh-keygen.

### CODE REVIEW

**Comment by systemcrash** [comment]
"Missing commit message. Please describe your changes"

**Comment by stokito** [comment]
Sure, I'll fix that. For now this is a draft PR to ask if it will be accepted. Please vote or tell what you think.

**Review by neheb — APPROVED** [review]

**Comment by neheb** [comment]
Seems great to be able to drop custom dropbearkey handling.

**Comment by neheb** [comment]
```
git grep dropbearkey | cut -d ':' -f 1 | sort -u
net/atlas-sw-probe/files/atlas.init
net/pppossh/README.md
```

sounds like dropbearkey can get removed actually...

**Comment by stokito** [comment]
The plain is:

1. Wait until the DB will start recognizing the `ssh-keygen` symlink after the PR merged https://github.com/mkj/dropbear/pull/265
2. Wait for the next DB release that will contain the changes. Then update DB in the OpenWrt feed. Today's DB version is 2022.83.
3. Merge the PR and the related https://github.com/openwrt/packages/pull/22861

**Comment by stokito** [comment]
There are two differences remains between dropbearkey and openssh ssh-keygen:
1. For the dropbearkey `-t` is required but for the ssh-keygen it will use RSA by default.
2. The ssh-keygen saves a pub key but for the dropbearkey not. So the additional command is needed `dropbearkey -y -f ~/.ssh/id_rsa | grep "^ssh- " > ~/.ssh/id_rsa.pub`

**Comment by systemcrash** [comment]
> There are two differences remains between dropbearkey and openssh ssh-keygen:
> 
>     1. For the dropbearkey `-t` is required but for the ssh-keygen it will use RSA by default.
> 

Depends what keytype you want to produce: in both cases therefore `-t` is necessary if you do not assume RSA. But yes, this is a difference.

>     2. The ssh-keygen saves a pub key but for the dropbearkey not. So the additional command is needed `dropbearkey -y -f ~/.ssh/id_rsa | grep "^ssh- " > ~/.ssh/id_rsa.pub`

Being explicit here is not a bad thing, esp since the openwrt device will use its new pub key to auth with remote devices.

That command by itself is incomplete. If I run it on my device (which has no `~/.ssh/id_rsa`), the pub key is empty (so a few more checks are required), and it potentially overwrites any pre-existing pubkey files. But yes, a difference.

Something like `xargs -r` (`	-r	Don't run command if input is empty`) should be of use here.

**Comment by stokito** [comment]
Ok, it looks like we can't make that much with the missing .pub files:
https://github.com/mkj/dropbear/pull/267

But make the rsa by default is easy:
https://github.com/mkj/dropbear/pull/268

**Comment by systemcrash** [comment]
Where is this today? Is it still waiting for a new dropbear version?

**Comment by stokito** [comment]
The Dropbear released the new version but the part with ssh-keygen is already added in patches:

https://github.com/openwrt/openwrt/blob/b5cde260487eae86db1661a53e5e5e0823936aab/package/network/services/dropbear/patches/018-dropbearkey-add-alias-to-ssh-keygen.patch

So we can merge it. I tested and it works.
Please merge the PR and corresponding https://github.com/openwrt/packages/pull/22861/files

**Comment by stokito** [comment]
@rockdrilla @rsalvaterra Please review

**Comment by rockdrilla** [comment]
LGTM

**Comment by stokito** [comment]
Thank you @Ansuel !

**Pull request merged**
